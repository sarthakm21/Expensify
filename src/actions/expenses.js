import db from "../firebase/firebase";

//ADD_EXPENSE
const addExpense = (expense) => ({
  type: "ADD_EXPENSE",
  expense,
});

const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0,
    } = expenseData;
    const expense = { description, note, amount, createdAt };
    return db
      .ref("expenses") //returning this so that the tests can be run on this
      .push(expense)
      .then((data) => {
        dispatch(
          addExpense({
            id: data.key,
            ...expense,
          })
        );
      })
      .catch((e) => console.log("Some error occured", e));
  };
};

//REMOVE_EXPENSE
const removeExpense = (id) => ({
  type: "REMOVE_EXPENSE",
  id,
});

const startRemoveExpense = (id) => {
  return (dispatch) => {
    db.ref(`expenses/${id}`)
      .remove()
      .then((data) => {
        dispatch(removeExpense(id));
      })
      .catch((e) => console.log("Some error occured", e));
  };
};

//EDIT_EXPENSE
const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates,
});

const startEditExpense = (id, updates) => {
  return (dispatch) => {
    db.ref(`expenses/${id}`)
      .update(updates)
      .then((data) => {
        dispatch(editExpense(id, updates));
      })
      .catch((e) => console.log("Some error occured", e));
  };
};

//SET EXPENSE
const setExpense = (expenses) => ({
  type: "SET_EXPENSE",
  expenses,
});

const startSetExpense = () => {
  return (dispatch) => {
    return db
      .ref("expenses")
      .once("value")
      .then((snapshot) => {
        const expenses = [];
        snapshot.forEach((child) => {
          expenses.push({ id: child.key, ...child.val() });
        });
        dispatch(setExpense(expenses));
      })
      .catch((e) => console.log(e));
  };
};

export {
  addExpense,
  removeExpense,
  editExpense,
  setExpense,
  startAddExpense,
  startEditExpense,
  startRemoveExpense,
  startSetExpense,
};
