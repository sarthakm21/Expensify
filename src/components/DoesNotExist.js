import React from "react";
const DoesNotExist = () => <div>404! The page requested does not exist.</div>;
export default DoesNotExist;
