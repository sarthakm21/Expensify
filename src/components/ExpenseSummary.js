import React from "react";
import expensesTotal from "../selectors/expenses-total";
import getVisibleExpenses from "../selectors/expenses";
import { connect } from "react-redux";
import numeral from "numeral";

export const ExpenseSummary = ({ totalExpenses, totalAmount }) => {
  //   const visibleExpenses = getVisibleExpenses(expenses, filters);
  //   const totalAmount = numeral(expensesTotal(visibleExpenses)).format("$0,0.00");
  //   const totalExpenses = visibleExpenses.length;
  return (
    <h2>
      Viewing {totalExpenses} expenses totalling to {totalAmount}
    </h2>
  );
};

const mapState = ({ expenses, filters }) => {
  const visibleExpenses = getVisibleExpenses(expenses, filters);
  const totalAmount = numeral(expensesTotal(visibleExpenses)).format("$0,0.00");
  const totalExpenses = visibleExpenses.length;
  return {
    totalExpenses,
    totalAmount,
  };
};

export default connect(mapState)(ExpenseSummary);
