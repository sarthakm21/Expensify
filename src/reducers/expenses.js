export default (state = [], action) => {
  switch (action.type) {
    case "ADD_EXPENSE":
      return [...state, action.expense];

    case "REMOVE_EXPENSE":
      return state.filter(({ id }) => id !== action.id);

    case "EDIT_EXPENSE":
      return state.map((expense) => {
        if (expense.id === action.id) {
          return {
            ...expense,
            ...action.updates,
          };
        } else return expense;
        //Object spread syntax does not work with many browsers so we need to add a babel plugin for this.
      });

    case "SET_EXPENSE":
      return action.expenses;

    default:
      return state;
  }
};
