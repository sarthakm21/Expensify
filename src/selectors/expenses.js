import moment from "moment";

const getVisibleExpenses = (expenses, { startDate, endDate, text, sortBy }) => {
  return expenses
    .filter((expense) => {
      const createdAtMoment = moment(expense.createdAt);
      const startDateMatches = startDate
        ? startDate.isSameOrBefore(createdAtMoment, "day")
        : true;

      const endDateMatches = endDate
        ? endDate.isSameOrAfter(createdAtMoment, "day")
        : true;

      const textMatches = expense.description
        .toLowerCase()
        .includes(text.toLowerCase());

      return startDateMatches && endDateMatches && textMatches;
    })
    .sort((a, b) => {
      if (sortBy === "date") {
        return moment(a.createdAt).isAfter(b.createdAt) ? -1 : 1;
      } else {
        return a.amount > b.amount ? -1 : 1;
      }
    });
};

export default getVisibleExpenses;
