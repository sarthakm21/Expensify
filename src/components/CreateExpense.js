import React from "react";
import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
import { addExpense } from "../actions/expenses";

export class CreateExpense extends React.Component {
  onSubmit = (expense) => {
    console.log(expense)
    this.props.addExpense(expense);
    this.props.history.push("/");
  }

  render() {
    return (
      <div>
        <div>
          This is the create element
          <ExpenseForm
            onSubmit={this.onSubmit}
            text="Add Expense"
          />
        </div>
      </div>
    );
  }
}

//aka mapDispatchToProps. Does the same thing to dispatch that mapStateToProps does to state
const mapDispatch = (dispatch) => ({
  addExpense: (expense) => dispatch(addExpense(expense))
})

export default connect(undefined, mapDispatch)(CreateExpense);
