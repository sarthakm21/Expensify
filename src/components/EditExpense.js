import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { startEditExpense, startRemoveExpense } from "../actions/expenses";

export class EditExpense extends React.Component {
  onSubmit = (expense) => {
    //Func. gets called by the submit event handler in the form which passes the new data as a parameter
    console.log(expense);
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push("/");
  };

  onRemove = () => {
    this.props.startRemoveExpense(this.props.expense.id); //Can also use props.match.params.id since both are same
    this.props.history.push("/");
  };
  render() {
    return (
      <div>
        <ExpenseForm
          expense={this.props.expense} //The expense received from mapStateToProps, i.e. the expense with an id same as the props.match.params.id
          onSubmit={this.onSubmit}
          text="Edit Expense"
        />
        <button onClick={this.onRemove}>Remove</button>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(
    (expense) => expense.id === props.match.params.id
  ),
});

const mapDispatch = (dispatch) => ({
  startRemoveExpense: (id) => dispatch(startRemoveExpense(id)),
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
});

export default connect(mapStateToProps, mapDispatch)(EditExpense);
