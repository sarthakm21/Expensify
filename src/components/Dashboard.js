import React from "react";
import ExpenseList from "./ExpenseList";
import ExpenseListFilters from "./ExpenseListFilters";
import ExpenseSummary from "./ExpenseSummary";

const Dashboard = () => (
  <div>
    <ExpenseSummary />
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);
export default Dashboard;
