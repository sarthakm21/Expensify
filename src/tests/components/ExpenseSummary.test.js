import React from "react";
import { ExpenseSummary } from "../../components/ExpenseSummary";
import expenses from "../fixtures/expenseData";
import { filters } from "../fixtures/filterData";
import { shallow } from "enzyme";

test("should render expense summary with single expense", () => {
  const wrapper = shallow(
    <ExpenseSummary totalExpenses={1} totalAmount={2031.32} />
  );
  expect(wrapper).toMatchSnapshot();
});

test("should render expense summary with multiple expenses", () => {
  const wrapper = shallow(
    <ExpenseSummary totalExpenses={4} totalAmount={209031.32} />
  );
  expect(wrapper).toMatchSnapshot();
});
