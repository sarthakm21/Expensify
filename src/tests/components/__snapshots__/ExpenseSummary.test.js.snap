// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense summary with multiple expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        4
      </span>
       expenses totalling to
       
      <span>
        209031.32
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render expense summary with single expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        1
      </span>
       expenses totalling to
       
      <span>
        2031.32
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
